lib/daru/maths/arithmetic/vector.rb

```diff
@@ -42,6 +42,30 @@ def round precision=0
           recode { |e| e.round(precision) unless e.nil? }
         end
 
+        # Add specified vector. Treat nil as 0.
+        #
+        # @example
+        #
+        #    v0 = Daru::Vector.new [1, 2, nil]
+        #    v1 = Daru::Vector.new [2, 1, 3]
+        #    irb> v0.add_skipnil v1
+        #    => #<Daru::Vector(3)>
+        #       0   3
+        #       1   3
+        #       2   3
+        #
+        def add_skipnil other
+          index = (@index.to_a | other.index.to_a)
+          elements = index.map do |idx|
+            this = self.index.include?(idx) ? self[idx] : nil
+            that = other.index.include?(idx) ? other[idx] : nil
+            this = 0 if this.nil?
+            that = 0 if that.nil?
+            this && that ? this + that : nil
+          end
+          Daru::Vector.new(elements, name: @name, index: index)
+        end
```

Comment:
I believe it is too verbose way. WDYT about this:
```ruby
# existing method, just update it a bit:
def v2v_binary operation, other, empty = nil
  index = (@index.to_a | other.index.to_a).sort

  elements = index.map do |idx|
    this = self.index.include?(idx) ? self[idx] : nil
    that = other.index.include?(idx) ? other[idx] : nil

    this && that ? this.send(operation, that) : empty
  end

  Daru::Vector.new(elements, name: @name, index: index)
end

# more poverful synonym for +
def add other, empty = nil
  binary_op :+, other, empty
end

# finally, in vector_sum:
...
vecs.inject(sum) { |memo, n| memo.add(self[n], skipnil ? 0 : nil) }
...

Reply:
@zverok an idea of clean `add` looks good. The suggested implementation will need some changes. In the above, `[1, 2, nil] add [1, 2,  3]` will produce `[1, 2, 0]`. To make this verbose I am thinking an approach similar to what you suggest but having a condition added to `v2v_binary`:

```ruby

            if (this.nil? || that.nil?) && skipnil
              this = 0 if this.nil?
              that = 0 if that.nil?
            end
```

Note the additional `(this.nil? || that.nil?) `. Reviewing the code I realized, that `nil add nil` for the vector should produce `nil`. Without this condition, it will produce `0`.

Will push a change along the lines suggested by you and the condition added.

Reply:
I was mistaken about `nil add nil`. 0 makes sense as its consistent with `DataFrame#vector_sum` which zeroes out the value.